src/qibolab/instruments/zhinst.py

```diff
@@ -659,22 +660,46 @@ def define_exp(self, qubits, couplers, exp_options, exp):
             self.experiment = exp
 
     def set_calibration(self, exp):
-        calib = lo.Calibration()
         if self.processed_sweeps:
             for ch in (
                 set(self.sequence.keys()) | self.processed_sweeps.channels_with_sweeps()
             ):
                 for param, sweep_param in self.processed_sweeps.sweeps_for_channel(ch):
                     if param is Parameter.frequency:
-                        calib[ch] = lo.SignalCalibration(
-                            oscillator=lo.Oscillator(
-                                frequency=sweep_param,
-                                modulation_type=lo.ModulationType.HARDWARE,
+                        exp.set_calibration(
+                            lo.Calibration(
+                                {
+                                    ch: lo.SignalCalibration(
+                                        oscillator=lo.Oscillator(
+                                            frequency=sweep_param,
+                                            modulation_type=lo.ModulationType.HARDWARE,
+                                        )
+                                    )
+                                }
                             )
                         )
+
+                    # for the remaining cases we can only achieve our goal by manipulating the instrument node directly
+                    channel_node_path = None
+                    logical_signal = self.signal_map[ch]
+                    for instrument in self.device_setup.instruments:
+                        for conn in instrument.connections:
+                            if conn.remote_path == logical_signal.path:
+                                channel_node_path = (
+                                    f"{instrument.address}/{conn.local_port}"
+                                )
+                                break
+                    if channel_node_path is None:
+                        raise RuntimeError(
+                            f"Could not find instrument node corresponding to {param.name} of channel {ch}"
+                        )
```

Comment:
This could be a tiny function on its own (just takes `self`, and returns `channel_node_path`), and you could call it even before `if param is Parameter.frequency:`.
It is true that that first `if` block does not require it, but it gets executed anyhow, and if it raises the error better to do it immediately.

Moreover, you could turn the various `if`s into `elif` (since the parameter is only one thing among those).
Using an `Enum` you could even avoid the `if`s, and just make a dictionary of functions accepting `exp, sweep_param, channel_node_path`. But this is a matter of taste.

Reply:
@alecandido I am playing around with code at the moment, trying to understand what works best for handling some limitations that I discovered yesterday. I will officially request review once I am done. I am not discouraging to review these intermediate commits though :D It will make the final review easier, I am just saying not to expect finalized nice code yet.